### PR TITLE
Multiplexer gains extra outputs and wraps out-of-range indices.

### PR DESCRIPTION
**Pull request by Gurkenglas**
## About The Pull Request

It took way too much boilerplate to wrap a counter, so I made the multiplexer do it. It now also includes the powers of a demultiplexer.

image

## Why It's Good For The Game

Reduce boilerplate - it's silly to read a spaghetti pile of arithmetic only to realize it counts 1,2,3,4,1,2,3,4,.... in the most compact way available. Complicated circuits that do things at least give hints by the components involved that do things.
It should also be easy to route data to one of multiple systems. Currently this can be done with the compare component, but it only routes signals, two ways at a time, while the data has to go everywhere. This way streamlines the spaghetti.

## Changelog
:cl:
expansion: The multiplexer component, now router, also demultiplexes across multiple outputs. Also, the indices now wrap instead of clamp.
/:cl:

### CODE REVIEW

**Comment by Gurkenglas** [comment]
https://github.com/tgstation/tgstation/compare/master...Gurkenglas:fdb36aa displays the changes without the rename, which should be cleaner to read.

File: code/modules/wiremod/components/utility/router.dm [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+		PORT_TYPE_LIST,
+		PORT_TYPE_ATOM,
+	)
+	options = component_options
+
+/obj/item/circuit_component/router/Initialize()
+	. = ..()
+	current_type = current_option
+	nin = add_input_port("Input", PORT_TYPE_NUMBER, default = 1)
+	nout = add_input_port("Output", PORT_TYPE_NUMBER, default = 1)
```
Comment (Watermelon914) [inline code/modules/wiremod/components/utility/router.dm:40]: These names are too vague and will be incredibly misleading and hard to reference.

**Review by Watermelon914 — COMMENTED** [review]
I think you should leave the multiplexer in and make this its own component, since the multiplexer is a lot more simple in design and this is more likely to overwhelm someone who doesn't know what they're doing, although they do have similar functions.

Additionally, if you can properly name the variables and properly set the description to be a lot more informative and easier to read, that'd also be a decent alternative.

File: code/modules/wiremod/components/utility/router.dm [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	)
+	options = component_options
+
+/obj/item/circuit_component/router/Initialize()
+	. = ..()
+	current_type = current_option
+	if(input_port_amount > 1)
+		which_input = add_input_port("'Which Input?'", PORT_TYPE_NUMBER, default = 1)
+	if(output_port_amount > 1)
+		which_output = add_input_port("'Which Output?'", PORT_TYPE_NUMBER, default = 1)
```
Comment (Watermelon914) [inline code/modules/wiremod/components/utility/router.dm:42]: Just rename these selectors instead. They shouldn't be questions.

File: code/modules/wiremod/components/utility/router.dm [hunk 3]
```diff
@@ -0,0 +1,77 @@
+/**
+ * # Router Component
+ *
+ * Writes one of multiple inputs to one of multiple outputs.
+ */
+/obj/item/circuit_component/router
+	display_name = "Router"
+	display_desc = "Writes an input of your choice to an output of your choice. If you set 'Which Input?' to any of ...-6,-2,2,6,10,... and 'Which Output?' to any of ...-5,-1,3,7,11,..., Input 2 will be written to Output 3."
```
Comment (Watermelon914) [inline code/modules/wiremod/components/utility/router.dm:8]: Don't need an example here. Just make sure to explain what it is

File: code/modules/wiremod/components/utility/router.dm [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+		output_selector = add_input_port("Output Selector", PORT_TYPE_NUMBER, default = 1)
+	ins = list()
+	for(var/port_id in 1 to input_port_amount)
+		ins += add_input_port(input_port_amount > 1 ? "Input [port_id]" : "Input", current_type)
+	outs = list()
+	for(var/port_id in 1 to output_port_amount)
+		outs += add_output_port(output_port_amount > 1 ? "Output [port_id]" : "Output", current_type)
+
+/obj/item/circuit_component/router/Destroy()
+	ins.Cut()
```
Comment (Watermelon914) [inline code/modules/wiremod/components/utility/router.dm:51]: you need to set the input_selector and output_selector to null here.

**Pull request merged**